File: Java/src/_0006_ZigZag_Conversion/Solution.java
package _0006_ZigZag_Conversion;

import java.util.ArrayList;
import java.util.List;

/**
 * Date: 2020/8/28 11:40
 * Content:
 */
public class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1)
            return s;

        List<StringBuilder> list = new ArrayList<>();
        for (int i = 0; i < Math.min(numRows, s.length()); i++) {
            list.add(new StringBuilder());
        }

        boolean down = false;
        int curRow = 0;
        for (char c : s.toCharArray()) {
            list.get(curRow).append(c);
            if (curRow == 0 || curRow == numRows-1)
                down = !down;
            curRow += (down ? 1 : -1);
        }

        StringBuilder res = new StringBuilder();
        for (StringBuilder row : list)
            res.append(row);
        return res.toString();
    }

    public static void main(String[] args) {
        String s = "LEETCODEISHIRING";
        int numRows = 3;
        System.out.println((new Solution()).convert(s, numRows));
    }
}

File: Python/_0006_ZigZag_Conversion/Solution.py
# coding=utf-8


# 执行用时：88 ms, 在所有 Python3 提交中击败了20.39%的用户
# 内存消耗：14.9 MB, 在所有 Python3 提交中击败了83.26%的用户
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        d = 2 * numRows - 2  # 公差
        res = ""
        for i in range(numRows):
            if i == 0 or i == numRows - 1:
                for j in range(i, len(s), d):
                    res += s[j]
            else:
                j = i; k = d - i
                while j < len(s) or k < len(s):
                    if j < len(s):
                        res += s[j]
                        j += d
                    if k < len(s):
                        res += s[k]
                        k += d
        return res


if __name__ == "__main__":
    print(Solution().convert("PAYPALISHIRING", 3))
